tensorflow/stream_executor/cuda/cusparse_stub.cc

```diff
@@ -53,14 +53,12 @@ cusparseStatus_t GetSymbolNotFoundError() {
 
 #if CUDA_VERSION < 10000
 #include "tensorflow/stream_executor/cuda/cusparse_9_0.inc"
-#elif CUDA_VERSION == 10000
+#elif CUDA_VERSION < 10010
 #include "tensorflow/stream_executor/cuda/cusparse_10_0.inc"
-#elif CUDA_VERSION == 10010
+#elif CUDA_VERSION < 10020
 #include "tensorflow/stream_executor/cuda/cusparse_10_1.inc"
-#elif CUDA_VERSION == 10020
+#elif CUDA_VERSION < 11000
 #include "tensorflow/stream_executor/cuda/cusparse_10_2.inc"
-#elif CUSPARSE_VER_MAJOR == 11 && CUSPARSE_VER_MINOR == 0
-#include "tensorflow/stream_executor/cuda/cusparse_11_0.inc"
 #else
-#error "We don't have a wrapper for this version."
+#include "tensorflow/stream_executor/cuda/cusparse_11_0.inc"
```

Comment:
The bit I didn't understand (from the PR and from our in person conversation) is why should this not be:

```
#if CUDA_VERSION < 10000
...
#elif CUDA_VERSION < 11000
...
#elif CUDA_VERSION < 12000
#include "tensorflow/stream_executor/cuda/cusparse_11_0.inc"
#else
#error
#endif
```

Why do we care about using CUDA 11 headers for CUDA 12?

Reply:
That would be OK too. My changes keeps the pattern a bit more consistent across all stubs, because a few ([cufft](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cufft_stub.cc), [curand](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/curand_stub.cc), and [cupti](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cupti_stub.cc)) continue to use the cuda 10 inc files even in cuda 11. In general assuming that a major version change will break APIs seems unnecessary.

Reply:
If using CUDA 11 inc files in CUDA 12 cannot result in a silent breakage then I'm ok with this pattern.  Otherwise I'd prefer keeping the explicit `#error` where possible.

Reply:
I don't think silent breakage is possible.

Case 1: Altered APIs (change arg ordering, etc.)
Because the toolkit headers are included in the stub modules (for example, [cusuparse](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cusparse_stub.cc#L16) and [cublas](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/cuda/cublas_stub.cc#L16-L18)) and overloading is not possible since the interfaces are declared `extern "C"`, any API changes will result in redefinition errors when the stubs are compiled.

Case 2: Removed APIs
Other modules using cuda lib APIs also include the original toolkit headers (rather than the generated `*.inc` files). So uses of removed APIs will result in undefined symbol errors when those modules are compiled.

Case 3: Use of newly added APIs
Users may add calls to new APIs introduced in CUDA 12 but forget to create an updated `*.inc` file. In this case, the code will compile fine, since the toolkit header includes the new API, but linking will fail, since the module is linked against the stub which does not yet expose this new API.

Reply:
> any API changes will result in redefinition errors when the stubs are compiled

I'm missing something here.  Let's say hypothetically cuDNN 9 swaps the two parameters to `cudnnGetTensorSizeInBytes`.  Won't we cast the function pointer to the wrong type [here](https://github.com/tensorflow/tensorflow/blob/c118bc4b5e1bb3a2760a026621d7a19053b0626c/tensorflow/stream_executor/cuda/cudnn_8_0.inc#L167)?

Reply:
If the arguments of `cudnnGetTensorSizeInBytes` were reversed, the cudnn v9 headers would forward declare the following. 
```
cudnnStatus_t CUDNNWINAPI cudnnGetTensorSizeInBytes(
    size_t *size, const cudnnTensorDescriptor_t tensorDesc);
```
Now building with this PR as it stands, `cudnn_stub.cc` would include that cudnn header and then the cudnn_8_0.inc file with this out of date function definition.
``` 
  cudnnStatus_t CUDNNWINAPI cudnnGetTensorSizeInBytes(
    const cudnnTensorDescriptor_t tensorDesc, size_t *size) {
  using FuncPtr =
      cudnnStatus_t(CUDNNWINAPI *)(const cudnnTensorDescriptor_t, size_t *);
  static auto func_ptr = LoadSymbol<FuncPtr>("cudnnGetTensorSizeInBytes");
  if (!func_ptr) return GetSymbolNotFoundError();
  return func_ptr(tensorDesc, size);
}
```
These functions are both declared `extern "C"`, so overloading the symbol `cudnnGetTensorSizeInBytes` is not allowed. The compiler will thus error out when it sees the same symbol with different argument lists.

Reply:
Ah, I missed the `extern "C"` at the top of the `.inc` file.  Thanks, LGTM then!